jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -34,89 +36,127 @@
 
     try (Timer ignored = new Timer(buildConfiguration.getBuildLogger(), DESCRIPTION)) {
-      RegistryClient registryClient =
-          new RegistryClient(
-              NonBlockingSteps.get(authenticatePullStep),
-              buildConfiguration.getBaseImageRegistry(),
-              buildConfiguration.getBaseImageRepository());
-
-      ManifestTemplate manifestTemplate =
-          registryClient.pullManifest(buildConfiguration.getBaseImageTag());
-
-      // TODO: Make schema version be enum.
-      switch (manifestTemplate.getSchemaVersion()) {
-        case 1:
-          V21ManifestTemplate v21ManifestTemplate = (V21ManifestTemplate) manifestTemplate;
-          return JsonToImageTranslator.toImage(v21ManifestTemplate);
-
-        case 2:
-          V22ManifestTemplate v22ManifestTemplate = (V22ManifestTemplate) manifestTemplate;
-          if (v22ManifestTemplate.getContainerConfiguration() == null
-              || v22ManifestTemplate.getContainerConfiguration().getDigest() == null) {
-            throw new UnknownManifestFormatException(
-                "Invalid container configuration in Docker V2.2 manifest: \n"
-                    + Blobs.writeToString(JsonTemplateMapper.toBlob(v22ManifestTemplate)));
-          }
-
-          ByteArrayOutputStream containerConfigurationOutputStream = new ByteArrayOutputStream();
-          registryClient.pullBlob(
-              v22ManifestTemplate.getContainerConfiguration().getDigest(),
-              containerConfigurationOutputStream);
-          String containerConfigurationString =
-              new String(containerConfigurationOutputStream.toByteArray(), StandardCharsets.UTF_8);
-
-          ContainerConfigurationTemplate containerConfigurationTemplate =
-              JsonTemplateMapper.readJson(
-                  containerConfigurationString, ContainerConfigurationTemplate.class);
-          return JsonToImageTranslator.toImage(v22ManifestTemplate, containerConfigurationTemplate);
+      // First, try with no credentials.
+      try {
+        return new Result(pullBaseImage(null), null);
+
+      } catch (RegistryUnauthorizedException ex) {
+        // If failed, then, retrieve base registry credentials and try with retrieved credentials.
+        // TODO: Refactor the logic in RetrieveRegistryCredentialsStep out to
+        // registry.credentials.RegistryCredentialsRetriever to avoid this direct executor hack.
+        ListeningExecutorService directExecutorService = MoreExecutors.newDirectExecutorService();
+        RetrieveRegistryCredentialsStep retrieveBaseRegistryCredentialsStep =
+            RetrieveRegistryCredentialsStep.forBaseImage(directExecutorService, buildConfiguration);
+
+        Authorization registryCredentials =
+            NonBlockingSteps.get(retrieveBaseRegistryCredentialsStep);
+        return new Result(pullBaseImage(registryCredentials), registryCredentials);
```

Comment:
Maybe `pullBaseImage` could just return a `Result` instead of having to wrap the returned image in a new one.

Reply:
Usually I think that'd be good, though here it would mean just calling the `new Result` in the returns in `pullBaseImage`, but `pullBaseImage` is intended to be a method that just pulls the base image using the credentials, whereas `Result` is the result of the step as a whole. I'll add a javadoc for this method.